django/contrib/admin/views/autocomplete.py

```diff
@@ -0,0 +1,167 @@
+import sys
+
+from django.core.exceptions import PermissionDenied
+from django.core.paginator import EmptyPage, Page, PageNotAnInteger, Paginator
+from django.db import router
+from django.http import Http404, JsonResponse
+from django.utils.encoding import force_text
+from django.utils.functional import cached_property
+from django.views.generic.list import BaseListView
+
+
+class AutocompletePage(Page):
+    def has_next(self):
+        """
+        Return true until queryset is smaller than the requested page size.
+
+        This avoids calling calculating the whole page count.
+        """
+        return self.object_list.count() >= self.paginator.per_page
+
+
+class AutocompletePaginator(Paginator):
+    """
+    A Django Paginator for big data sets.
+
+    Because users mostly use the first pages, it doesn't matter
+    if the page count is correct or not, as long as it's high enough.
+    Count() queries can take time, because they want to be exact.
+    So this paginator just doesn't count.
+    """
+
+    def _check_object_list_is_ordered(self):
+        """
+        Remove ordering and the corresponding warning.
+
+        Ordering is very slow, and we are doing a search here anyways.
+        There is not big advantage ordering the result set, especially
```

Comment:
I decided to drop the whole custom paginator. I will use the `ModelAdmin.paginator` class. If someone desires to change the paginators behavior to speed up things, it will be possible.
I guess it's a lot cleaner this way.